File: webapp/ruby/lib/notifier.rb
require 'webpush'
require 'xsuportal/resources/notification_pb'

module Xsuportal
  class Notifier
    WEBPUSH_VAPID_PRIVATE_KEY_PATH = '../vapid_private.pem'
    WEBPUSH_SUBJECT = 'xsuportal@example.com'

    def initialize(db)
      @db = db
    end

    def vapid_key
      @vapid_key ||= begin
        if File.exist?(WEBPUSH_VAPID_PRIVATE_KEY_PATH)
          private_key = File.read(WEBPUSH_VAPID_PRIVATE_KEY_PATH)
          Webpush::VapidKey.from_pem(private_key)
        else
          nil
        end
      end
    end

    def notify_clarification_answered(clar, updated: false)
      contestants = nil
      if clar[:disclosed]
        contestants = db.query('SELECT `id`, `team_id` FROM `contestants` WHERE `team_id` IS NOT NULL')
      else
        contestants = db.xquery(
          'SELECT `id`, `team_id` FROM `contestants` WHERE `team_id` = ?',
          clar[:team_id],
        )
      end

      contestants.each do |contestant|
        notification_pb = Proto::Resources::Notification.new(
          content_clarification: Proto::Resources::Notification::ClarificationMessage.new(
            clarification_id: clar[:id],
            owned: clar[:team_id] == contestant[:team_id],
            updated: updated,
          )
        )
        notification = notify(notification_pb, contestant[:id])
        if vapid_key
          notification_pb.id = notification[:id]
          notification_pb.created_at = notification[:created_at]
          # TODO: Web Push IIKANJI NI SHITE
        end
      end
    end

    def notify_benchmark_job_finished(job)
      contestants = db.xquery(
        'SELECT `id`, `team_id` FROM `contestants` WHERE `team_id` = ?',
        job[:team_id],
      )

      contestants.each do |contestant|
        notification_pb = Proto::Resources::Notification.new(
          content_benchmark_job: Proto::Resources::Notification::BenchmarkJobMessage.new(
            benchmark_job_id: job[:id],
          )
        )
        notification = notify(notification_pb, contestant[:id])
        if vapid_key
          notification_pb.id = notification[:id]
          notification_pb.created_at = notification[:created_at]
          # TODO: Web Push IIKANJI NI SHITE
        end
      end
    end

    private
    def db
      @db
    end

    def notify(notification, contestant_id)
      encoded_message = [Proto::Resources::Notification.encode(notification)].pack('m0')
      db.xquery(
        'INSERT INTO `notifications` (`contestant_id`, `encoded_message`, `read`, `created_at`, `updated_at`) VALUES (?, ?, FALSE, NOW(6), NOW(6))',
        contestant_id,
        encoded_message,
      )
      notification = db.query('SELECT * FROM `notifications` WHERE `id` = LAST_INSERT_ID() LIMIT 1').first
      notification
    end
  end
end

File: webapp/nodejs/src/notifier.ts
import fs from 'fs';
import webpush from 'web-push';
import sshpk from 'sshpk';
import urlBase64 from 'urlsafe-base64';
import util from 'util';

import type { PoolConnection } from 'promise-mysql';
import { Notification } from '../proto/xsuportal/resources/notification_pb';
import { convertDateToTimestamp } from './app';

const sleep = util.promisify(setTimeout);
export class Notifier {
  static WEBPUSH_VAPID_PRIVATE_KEY_PATH = '../vapid_private.pem';
  static WEBPUSH_SUBJECT = 'xsuportal@example.com';
  static VAPIDKey: webpush.VapidKeys;

  constructor() {
    this.getVAPIDKey()
  }

  getVAPIDKey() {
    if(Notifier.VAPIDKey) return Notifier.VAPIDKey;
    if(!fs.existsSync(Notifier.WEBPUSH_VAPID_PRIVATE_KEY_PATH)) return null;
    const pri = sshpk.parsePrivateKey(fs.readFileSync(Notifier.WEBPUSH_VAPID_PRIVATE_KEY_PATH), "pem");
    const pub = pri.toPublic();
    const privateKey = urlBase64.encode((pri as any).part.d.data)
    const publicKey = urlBase64.encode((pub as any).part.Q.data)
    webpush.setVapidDetails(`mailto:${Notifier.WEBPUSH_SUBJECT}`, publicKey, privateKey);
    Notifier.VAPIDKey = { privateKey, publicKey }
    return Notifier.VAPIDKey;
  }

  async notifyClarificationAnswered(clar: NonNullable<any>, db: PoolConnection, updated = false) {
    const contestants = await db.query(
      clar.disclosed
        ? 'SELECT `id`, `team_id` FROM `contestants` WHERE `team_id` IS NOT NULL'
        : 'SELECT `id`, `team_id` FROM `contestants` WHERE `team_id` = ?',
      [clar.team_id]
    );

      for (const contestant of contestants) {
        const clarificationMessage = new Notification.ClarificationMessage();
        clarificationMessage.setClarificationId(clar.id);
        clarificationMessage.setOwned(clar.team_id === contestant.team_id);
        clarificationMessage.setUpdated(updated);
        const notification = new Notification();
        notification.setContentClarification(clarificationMessage);
        const inserted = await this.notify(notification, contestant.id, db);
        if (inserted && Notifier.VAPIDKey) {
          notification.setId(inserted.id);
          notification.setCreatedAt(convertDateToTimestamp(inserted.created_at));
          // TODO Web Push IIKANJINI SHITE
        }
      }
  }

  async notifyBenchmarkJobFinished(job, db: PoolConnection) {
    const contestants = await db.query(
      'SELECT `id`, `team_id` FROM `contestants` WHERE `team_id` = ?',
      [job.team_id]
    );

    for (const contestant of contestants) {
      const benchmarkJobMessage = new Notification.BenchmarkJobMessage();
      benchmarkJobMessage.setBenchmarkJobId(job.id);
      const notification = new Notification();
      notification.setContentBenchmarkJob(benchmarkJobMessage);
      const inserted = await this.notify(notification, contestant.id, db);
      if (inserted && Notifier.VAPIDKey) {
        notification.setId(inserted.id);
        notification.setCreatedAt(convertDateToTimestamp(inserted.created_at));
        // TODO Web Push IIKANJINI SHITE
      }
    }
  }

  async notify(notification: Notification, contestantId, db: PoolConnection) {
    const encodedMessage = Buffer.from(notification.serializeBinary()).toString('base64');
    await db.query(
      'INSERT INTO `notifications` (`contestant_id`, `encoded_message`, `read`, `created_at`, `updated_at`) VALUES (?, ?, FALSE, NOW(6), NOW(6))',
      [contestantId, encodedMessage]
    );
    let [n] = await db.query('SELECT * FROM `notifications` WHERE `id` = LAST_INSERT_ID() LIMIT 1');
    return n
  }
}
